client/src/components/room/streamGate.tsx

```diff
@@ -0,0 +1,136 @@
+ *  */
+const StreamGate: React.FunctionComponent<{
+  children: any;
+  roomID: string;
+  streamPlaybackID?: string;
+  anonymousOnly: boolean;
+}> = ({ children, roomID, streamPlaybackID, anonymousOnly }) => {
+  const isAdmin = useAdminStore(useCallback((s) => s.isAdmin, []));
+
+  const [state, setState] = useState<{
+    client: StreamVideoClient;
+    call: Call;
+    rtmpsDetails: RtmpsDetails | null;
+  }>();
+
+  const useAdmin = !anonymousOnly && isAdmin;
+
+  useEffect(() => {
+    getClient(roomID, useAdmin).then(([myClient, rtmpsDetails]) => {
+      if (!streamPlaybackID) {
+        // Technically, streamPlaybackID may be indirectly set in the middle of getClient()
+        // so there is a race condition here. If this happens, we will call getClient twice,
+        // and that's okay.
+        return;
+      }
+
+      const myCall = myClient.call("livestream", streamPlaybackID);
+
+      myCall
+        .get()
+        .then(() => {
+          logInfo("Setting state for call ID " + streamPlaybackID);
+          setState({ client: myClient, call: myCall, rtmpsDetails });
+        })
+        .catch((e) => logError("Failed to retrieve call details", e));
+    });
+  }, [roomID, streamPlaybackID, useAdmin]);
+
+  useEffect(() => {
+    if (!state) {
+      return;
+    }
+
+    logInfo("....JOINING CALL....");
+
+    // Ensure camera is disabled by default on page load.
+    state.call.camera.disable();
+
+    state.call
+      .join()
+      .then(() => {
+        logInfo("Joined call from StreamGate");
+      })
+      .catch((e) => {
+        logError("Failed to join call", e);
+      });
+
+    return () => {
+      state.call.leave().catch((e) => {
+        logError("Failed to leave call", e);
+      });
+    };
+  }, [state]);
+
+  if (!state || !state.client || !state.call) return null;
+
+  return (
+    <StreamVideo client={state.client}>
+      <StreamCall call={state.call}>{children(state.rtmpsDetails)}</StreamCall>
+    </StreamVideo>
+  );
+};
+
+/** Retrieves the Stream video client as an admin or anonymous viewer. */
+const getClient = async (
+  roomId: string,
+  useAdmin: boolean,
+): Promise<[StreamVideoClient, null | RtmpsDetails]> => {
+  if (useAdmin) {
+    /**
+     * After we have verified the user as a room admin, we can make calls using secrets retrieved from the server and
+     * not worry too much about those credentials being viewable in the console.
+     */
+    console.log("using admin credentials for stream player");
+    const creds = await getStreamAdminCredentials(roomId);
+    const apiKey = process.env.NEXT_PUBLIC_STREAM_API_KEY!;
+
+    return [
+      new StreamVideoClient({
+        user: { id: creds.userId, name: "Admin" },
+        apiKey: apiKey,
+        token: creds.token,
+      }),
+      {
+        rtmpAddress: creds.call.rtmpAddress,
+        rtmpStreamKey: creds.call.rtmpStreamKey,
+      },
+    ];
+  } else {
+    console.log("using public credentials for stream player");
```

Comment:
Just to confirm, we need to let anonymous users join the call because we set it in StreamVideo + StreamCall, and this is NECESSARY to (later, in streamPlayer) correctly have the video work?

Just trying to understand because it was much simpler with mux. There was no "call" a public user had to join, as long as the streamPlaybackID was present, and the server correctly knew the stream was "active", one could publically watch a stream without worrying about "joining a call" or whatever.

Reply:
I got further confused by this, because streamGate is only within room. Not with index... but the video player seems to work on index.tsx. So again, wondering if you actually need anonymous folks to join the call?

Reply:
The streamGate is used in both the room component and the domRing component (which is why the index page works!) I agree it's a little unintuitive that the Stream stuff is centered around joining and leaving "calls", it seems like it was more designed for zoom-like calls than for livestreams.

 But yes -- anonymous users do need to "join" the call in order for them to receive the audio/video. The problem I was running into before was this:

1) Visit the /soft room before the livestream begins
2) Anonymous user could not join the "backstage" call
3) Streamer activates livestream
4) Since the anonymous user is not in the call, they do not receive audio/video. When they refresh the page, it joins the call and they receive A/V.

Reply:
Ahhh I see it in DomRing! Cool! Onwards!

Again, not blocking but do you think there are any performance implications of having many streams "joined" by the user? I'm asking because for our next stream, we may have up to 50 folks on the home page. I am imagining that we would no longer have the videos always displayed in this case, because 50 is way too many to show. Instead, we'd display (somehow) the video for each streamer on hover. 

In this case I see three options:
1. Simply don't have the video at all on home.
2. Be okay with all the joins happening on page open, but then only show video on hover
3. Have the stream gate activated "on hover"

Reply:
That's a good point. I imagine there would be performance implications (even without joining calls, playing 50 videos at once seems like it would take its toll.) 

I think (3) would be a good option to explore for the homepage, so the useEffect for joining/leaving the call would activate on hover instead.